libs/platform/ebpf_epoch.c

```diff
@@ -4,73 +4,53 @@
 #include "ebpf_epoch.h"
 
 // Brief summary of how epoch tracking works.
-// Each free operation increments the epoch and the freed memory is stamped with
-// that epoch.
+// Each free operation increments the _ebpf_current_epoch, the freed memory is stamped with that epoch, and the
+// memory is inserted into a per-CPU free list.
 //
-// Each block of code that accesses epoch freed memory wraps access in calls to
-// ebpf_epoch_enter/ebpf_epoch_exit.
+// Each block of code that accesses epoch freed memory wraps access in calls to ebpf_epoch_enter/ebpf_epoch_exit.
 //
-// Epoch tracking is handled differently for preemptible vs non-preemptible
-// invocations.
+// The per-CPU state is protected by a single per-CPU lock.
 //
-// Non-preemptible invocations behavior:
-// 1) During ebpf_epoch_enter and ebpf_epoch_exit the value of _ebpf_current_epoch is captured in the
-// _ebpf_epoch_cpu_table[cpu_id].non_preemptible_epoch field.
-// 2) This field is read/written without explicit synchronization and can be old.
-// 3) _ebpf_epoch_update_cpu_entry periodically updates this on idle CPUs.
+// ebpf_epoch_enter:
+// The current epoch is recorded in either a per thread (preemptible) or per CPU (non-preemtpible) ebpf_epoch_state_t
+// and it is marked as active.
 //
-// Preemptible invocations behavior:
-// 1) During ebpf_epoch_enter the value of _ebpf_current_epoch is captured and stored in the
-// _ebpf_epoch_cpu_table[cpu_id].thread_table[thread_id].entry_epoch field.
-// 2) During ebpf_epoch_exit the value of _ebpf_current_epoch is captured and stored in the
-// _ebpf_epoch_cpu_table[cpu_id].thread_table[thread_id].exit_epoch field.
-// 3) The _ebpf_epoch_cpu_table[cpu_id].thread_table is protected by _ebpf_epoch_cpu_table[cpu_id].thread_table_lock.
-// 4) If and only if entry_epoch > exit_epoch then the thread is actively executing between calls to ebpf_epoch_enter
-// and ebpf_epoch_exit.
+// ebpf_epoch_exit:
+// First:
+// The current epoch is recorded in either a per thread or per CPU ebpf_epoch_state_t and it is marked as inactive.
+// For the per-thread case, the current CPU's per-thread table is first checked for an active epoch record. If not
+// found then each CPU's per-thread table is checked until the active record is found. This deals with the case where
+// a thread may switch CPU's between enter and exit.
```

Comment:
```suggestion
// For the per-thread case, the current CPUs per-thread table is first checked for an active epoch record. If not
// found then each CPUs per-thread table is checked until the active record is found. This deals with the case where
// a thread may switch CPUs between enter and exit.
```
(nit) see https://abbreviations.yourdictionary.com/simple-rules-for-plural-acronyms-and-abbreviations.html and similar other sources